src/main.cpp

```diff
@@ -3511,10 +3511,19 @@ std::vector<unsigned char> GenerateCoinbaseCommitment(CBlock& block, const CBloc
             out.scriptPubKey[5] = 0xed;
             memcpy(&out.scriptPubKey[6], witnessroot.begin(), 32);
             commitment = std::vector<unsigned char>(out.scriptPubKey.begin(), out.scriptPubKey.end());
-            const_cast<std::vector<CTxOut>*>(&block.vtx[0].vout)->push_back(out);
-            block.vtx[0].UpdateHash();
+            mcoinbase.vout.emplace_back(std::move(out));
         }
     }
+
+    // We need a full copy of vtx to modify the coinbase.
+    std::vector<CTransaction> vtx;
+    vtx.reserve(block.vtx.size());
+    vtx.emplace_back(std::move(mcoinbase));
+    for (size_t i = 1; i < block.vtx.size(); i++) {
+        vtx.emplace_back(block.vtx[i]);
```

Comment:
If CTransaction is enitely immutable, a move constructor would just be identical to a copy constructor, so no need. We could leave a move constructor that only moves the (mutable) witness field, but that would get more complicated with #8589 (which moves the witness data inside CTxIn). I think a better (but follow-up) change would be changing CBlock::vtx to a `std::vector<std::shared_ptr<CTransaction>>`.